metals/src/main/scala/scala/meta/internal/metals/CodeActionProvider.scala

```diff
@@ -2,27 +2,60 @@ package scala.meta.internal.metals
 
 import org.eclipse.{lsp4j => l}
 import scala.meta.pc.CancelToken
+import scala.meta.internal.mtags.Semanticdbs
 import scala.concurrent.ExecutionContext
 import scala.concurrent.Future
+import scala.collection.JavaConverters._
 
 final class CodeActionProvider(
-    compilers: Compilers
+    compilers: Compilers,
+    trees: Trees,
+    buffers: Buffers,
+    semanticdbs: Semanticdbs,
+    symbolSearch: MetalsSymbolSearch,
+    definitionProvider: DefinitionProvider
 ) {
 
   def codeActions(
       params: l.CodeActionParams,
       token: CancelToken
   )(implicit ec: ExecutionContext): Future[Seq[l.CodeAction]] = {
 
-    for {
-      quickFixes <- Future
-        .sequence(
-          List(QuickFix.ImportMissingSymbol)
-            .map(_.contribute(params, compilers, token))
+    def quickfixes =
+      List(QuickFix.ImportMissingSymbol)
+        .map(_.contribute(params, compilers, token))
+
+    def refactorings =
```

Comment:
if we go with my proposal below, we can unify these `contribute` calls can go together

Reply:
I agree it's probably a good idea to get rid of the `QuickFix` and `Refactoring` traits and just have a single `CodeAction` trait to cover both.

But we should still avoid doing unnecessary work when the client sets the `only` param. e.g. if the client requests only quick fixes, we shouldn't waste cycles on calculating all possible refactorings. So we could either keep separate manually-curated lists, like I did here, or the `CodeAction` trait could have a `def kind: CodeActionKind` and we could filter based on that before we call `contribute`. I think the latter is probably cleaner.

/cc @tgodzik

Reply:
Ah, absolutely. I didn't think of `only` so I guess you can ignore my comment and we should invoke `contribute` only after we check that parameter.